datafusion/optimizer/src/simplify_expressions/expr_simplifier.rs

```diff
@@ -1681,9 +1679,15 @@ impl<'a, S: SimplifyInfo> TreeNodeRewriter for Simplifier<'a, S> {
     }
 }
 
-fn has_common_conjunction(lhs: &Expr, rhs: &Expr) -> bool {
+fn has_common_conjunction(lhs: &Expr, rhs: &Expr) -> Result<bool, DataFusionError> {
     let lhs: HashSet<&Expr> = iter_conjunction(lhs).collect();
-    iter_conjunction(rhs).any(|e| lhs.contains(&e))
+    iter_conjunction(rhs).try_fold(false, |acc, e| {
```

Comment:
This still looks a bit ugly. I mean, we just look for the first non-volatile expression and don't want to iterate over all expressions.

Actually, I think I made a mistake when I added `is_volatile()` with return type `Result<bool>`:
https://github.com/apache/datafusion/blob/e22d23113f549a90483bbf161add84eba08510fd/datafusion/expr/src/expr.rs#L1581-L1583
As the closure we provide to `exists()` can't return an error it would be safe to unwrap the result and return only `bool`.

@alamb, do you think we can fix `Expr::is_volatile()`, it would be an API breaking change though, to make it simpler and so we can keep using `any()` here?

Reply:
I have refactored it, Thanks for your review

Reply:
Looks good to me, thanks for the fix.

Maybe we can consider changing the return type of `Expr::is_volatile()` in a follow-up.

Reply:
> Maybe we can consider changing the return type of `Expr::is_volatile()` in a follow-up PR.

I think this PR would be much cleaner if we fixed the signature of that first. But I guess the important thing is that we do fix it before it spreads keeps spreading.

Reply:
I changed the signature in https://github.com/apache/datafusion/pull/13206 which caused a logical conflict in this PR (failed to compile) and then I updated this PR to use the new signature

As @eejbyfeldt  predicated, the PR became much cleaner